src/core/credGrainView.js

```diff
@@ -54,77 +60,133 @@ export type ParticipantsOptions = {|
+          if (!grain) grain = G.ZERO;
+          if (
+            interval.startTimeMs < allocationReceipt.credTimestampMs &&
+            allocationReceipt.credTimestampMs <= interval.endTimeMs
+          )
+            grain = G.add(grain, allocationReceipt.grainReceipt.amount);
+        }) || G.ZERO;
+      grainEarnedPerInterval.push(grainEarnedThisInterval);
+    }
+    return grainEarnedPerInterval;
   }
 
   withTimeScope(
-    startTimeMs: number,
-    endTimeMs: number
+    startTimeMs: TimestampMs,
+    endTimeMs: TimestampMs
   ): TimeScopedCredGrainView {
     return new TimeScopedCredGrainView(
-      this._credGraph,
-      this._ledger,
+      this._participants,
+      this._intervals,
       startTimeMs,
       endTimeMs
     );
   }
 
   intervals(): IntervalSequence {
-    return this._defaultTimeScope.intervals();
+    return cloneDeep(this._intervals);
   }
 
-  participant(id: IdentityId): ParticipantCredGrain | null {
-    return this._defaultTimeScope.participant(id);
+  participants(): $ReadOnlyArray<ParticipantCredGrain> {
+    return this._participants;
   }
 
-  participants(
-    options: ParticipantsOptions
-  ): $ReadOnlyArray<ParticipantCredGrain> {
-    return this._defaultTimeScope.participants(options);
+  participant(id: IdentityId): ParticipantCredGrain | null {
+    const _ = id;
+    throw new Error("not yet implemented");
   }
 }
 
 export class TimeScopedCredGrainView {
-  _credGraph: CredGraph;
-  _ledger: Ledger;
-  _startTimeMs: TimestampMs;
-  _endTimeMs: TimestampMs;
+  _participants: Array<ParticipantCredGrain>;
+  _intervals: IntervalSequence;
 
   constructor(
-    credGraph: CredGraph,
-    ledger: Ledger,
+    originalParticipants: $ReadOnlyArray<ParticipantCredGrain>,
+    originalIntervals: IntervalSequence,
     startTimeMs: TimestampMs,
     endTimeMs: TimestampMs
   ) {
-    this._credGraph = credGraph;
-    this._ledger = ledger;
-    this._startTimeMs = startTimeMs;
-    this._endTimeMs = endTimeMs;
+    let inclusiveStartIndex = originalIntervals.findIndex(
+      (interval) => startTimeMs <= interval.startTimeMs
+    );
+    if (inclusiveStartIndex === -1)
+      inclusiveStartIndex = originalIntervals.length;
+
+    const exclusiveEndIndex =
+      findLastIndex(
+        originalIntervals,
+        (interval) => interval.endTimeMs <= endTimeMs
+      ) + 1;
+
+    this._intervals = intervalSequence(
+      originalIntervals.slice(inclusiveStartIndex, exclusiveEndIndex)
+    );
+    this._participants = originalParticipants.map((participant) => {
+      const credPerInterval = participant.credPerInterval.slice(
+        inclusiveStartIndex,
+        exclusiveEndIndex
+      );
+      const grainEarnedPerInterval = participant.grainEarnedPerInterval.slice(
+        inclusiveStartIndex,
+        exclusiveEndIndex
+      );
+      return {
+        identity: participant.identity,
+        cred: credPerInterval.reduce((a, b) => a + b),
```

Comment:
Reduce should have an initial value of `0`. Otherwise it will error when the array is empty:

```js
> [].reduce((a,b) => a + b)
Uncaught TypeError: Reduce of empty array with no initial value
```

Reply:
accepted